codes/c/chapter_tree/avl_tree.c

```diff
@@ -0,0 +1,260 @@
+    if (bf > 1) {
+        if (balanceFactor(node->left) >= 0) {
+            // 右旋
+            return rightRotate(node);
+        } else {
+            // 先左旋后右旋
+            node->left = leftRotate(node->left);
+            return rightRotate(node);
+        }
+    }
+    // 右偏树
+    if (bf < -1) {
+        if (balanceFactor(node->right) <= 0) {
+            // 左旋
+            return leftRotate(node);
+        } else {
+            // 先右旋后左旋
+            node->right = rightRotate(node->right);
+            return leftRotate(node);
+        }
+    }
+    // 平衡树，无需旋转，直接返回
+    return node;
+}
+
+/* 递归插入结点（辅助函数） */
+TreeNode *insertHelper(TreeNode *node, int val) {
+    if (node == NULL) {
+        return newTreeNode(val);
+    }
+    /* 1. 查找插入位置，并插入结点 */
+    if (val < node->val) {
+        node->left = insertHelper(node->left, val);
+    } else if (val > node->val) {
+        node->right = insertHelper(node->right, val);
+    } else {
+        // 重复结点不插入，直接返回
+        return node;
+    }
+    // 更新结点高度
+    updateHeight(node);
+    /* 2. 执行旋转操作，使该子树重新恢复平衡 */
+    node = rotate(node);
+    // 返回子树的根节点
+    return node;
+}
+
+
+/* 插入结点 */
+TreeNode *insert(avlTree *tree, int val) {
+    tree->root = insertHelper(tree->root, val);
+    return tree->root;
+}
+
+/* 获取中序遍历中的下一个结点（仅适用于 root 有左子结点的情况） */
+TreeNode *getInOrderNext(TreeNode *node) {
+    if (node == NULL) {
+        return node;
+    }
+    // 循环访问左子结点，直到叶结点时为最小结点，跳出
+    while (node->left != NULL) {
+        node = node->left;
+    }
+    return node;
+}
+
+/* 递归删除结点（辅助函数） */
+TreeNode *removeHelper(TreeNode *node, int val) {
+    TreeNode *child, *grandChild, *temp;
+    if (node == NULL) {
+        return NULL;
+    }
+    /* 1. 查找结点，并删除之 */
+    if (val < node->val) {
+        node->left = removeHelper(node->left, val);
+    } else if (val > node->val) {
+        node->right = removeHelper(node->right, val);
+    } else {
+        if (node->left == NULL || node->right == NULL) {
+            child = node->left;
+            if (node->right != NULL) {
+                child = node->right;
+            }
+            // 子结点数量 = 0 ，直接删除 node 并返回
+            if (child == NULL) {
+                return NULL;
+            } else {
+                // 子结点数量 = 1 ，直接删除 node
+                node = child;
+            }
+        } else {
+            // 子结点数量 = 2 ，则将中序遍历的下个结点删除，并用该结点替换当前结点
+            temp = getInOrderNext(node->right);
+            node->right = removeHelper(node->right, temp->val);
+            node->val = temp->val;
+        }
+    }
+    // 更新结点高度
+    updateHeight(node);
+    /* 2. 执行旋转操作，使该子树重新恢复平衡 */
+    node = rotate(node);
+    // 返回子树的根节点
+    return node;
+}
+
+/* 删除结点 */
+TreeNode *removeNode(avlTree *tree, int val) {
+    TreeNode *root = removeHelper(tree->root, val);
+    return root;
+}
+
+/* 查找结点 */
+TreeNode *search(avlTree *tree, int val) {
+    TreeNode *cur = tree->root;
+    // 循环查找，越过叶结点后跳出
+    while (cur != NULL) {
+        // 目标结点在 root 的右子树中
```

Comment:
Move it to line 206 to match the location of the other two comments.

Reply:
Done.